test/testHelpers.hpp

```diff
@@ -527,199 +527,139 @@ af::dim4 unravelIdx(uint idx, af::array arr) {
     }
 
+    return out;
 }
 
 template<typename T>
 ::testing::AssertionResult elemWiseEq(std::string aName, std::string bName,
-                                      af::array a, af::array b,
-                                      FloatTag) {
-    uint nElems = a.elements();
-    af::dim4 arrDims = a.dims();
-
-    std::vector<T> hA(nElems);
-    a.host(hA.data());
-
-    std::vector<T> hB(nElems);
-    b.host(hB.data());
-
+                                      std::vector<T>& a, af::dim4 aDims,
+                                      std::vector<T>& b, af::dim4 bDims,
+                                      float maxAbsDiff, IntegerTag) {
     typedef typename std::vector<T>::iterator iter;
-    std::pair<iter, iter> mismatches = std::mismatch(hA.begin(), hA.end(), hB.begin());
+    std::pair<iter, iter> mismatches = std::mismatch(a.begin(), a.end(), b.begin());
     iter aItr = mismatches.first;
     iter bItr = mismatches.second;
 
-    if (aItr == hA.end()) {
+    if (bItr == b.end()) {
         return ::testing::AssertionSuccess();
     } else {
-        int idx = std::distance(hA.begin(), aItr);
-        af::dim4 coords = unravelIdx(idx, a.dims(), af::getStrides(a));
-
-        return ::testing::AssertionFailure() << "VALUE DIFFERS at ("
-                                             << coords[0] << ", " << coords[1] << ", "
-                                             << coords[2] << ", " << coords[3] << "): "
-                                             << aName << "(" << hA[idx] << "), "
-                                             << bName << "(" << hB[idx] << ")";
+        int idx = std::distance(b.begin(), bItr);
+        af::dim4 bStrides = calcStrides(bDims);
+        af::dim4 coords = unravelIdx(idx, bDims, bStrides);
+
+        return ::testing::AssertionFailure() << "VALUE DIFFERS:\n"
+                                             << " at ([" << coords << "]):\n"
+                                             << aName << "(" << a[idx] << ")\n"
+                                             << bName << "(" << b[idx] << ")";
     }
 }
 
 template<typename T>
-::testing::AssertionResult elemWiseEq(std::string hA_name, std::string bName,
-                                      std::vector<T>& hA, af::dim4 aDims, af::array b,
-                                      IntegerTag) {
-    std::vector<T> hB(b.elements());
-    b.host(hB.data());
-
+::testing::AssertionResult elemWiseEq(std::string aName, std::string bName,
+                                      std::vector<T>& a, af::dim4 aDims,
+                                      std::vector<T>& b, af::dim4 bDims,
+                                      float maxAbsDiff, FloatTag) {
     typedef typename std::vector<T>::iterator iter;
-    std::pair<iter, iter> mismatches = std::mismatch(hA.begin(), hA.end(), hB.begin());
+    // TODO(mark): Modify equality for float
+    std::pair<iter, iter> mismatches = std::mismatch(a.begin(), a.end(), b.begin());
```

Comment:
This can be moved into the else branch since the results aren't used in the true block.